File: p2p-java/src/Node.java
//import com.sun.security.ntlm.Server;

import java.io.*;
import java.net.*;
import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.ReentrantLock;

public class Node implements Runnable {
    private String host;
    private int port;
    private ServerSocket ServerSocket;
    private int terminate_flag;

    public Node(String host, int port) throws IOException{
        this.host = host;
        this.port = port;
        this.terminate_flag = 0;
        InetAddress addr = InetAddress.getByName(host);
        this.ServerSocket = new ServerSocket(port,50, addr);
        this.ServerSocket.setSoTimeout(10*1000);
        System.out.println("Created Node at IP: " + this.host + ", port: " + this.port);
    }

    public static void main(String[] args) throws IOException, InterruptedException {
        //Here we will try the functionality associated with the Node class
        //Creation of Node
        Node node = new Node("127.0.0.1", 18080);
        Thread t = new Thread(node);
        t.start();

        //Node node2 = new Node("127.0.0.1", 18080);
        //node2.send_message("HOLA", "127.0.0.1",8080);

        TimeUnit.SECONDS.sleep(1);

        node.stop_node();
        //node2.stop_node();
    }

    @Override
    public void run() {
        /**
         * run: Loop that waits for incoming socket connections and deals with them accordingly
         *
         */
        while(terminate_flag == 0){
            System.out.println("Node: Waiting for incoming connections...");
            try {
                Socket clientSocket = this.ServerSocket.accept();
                BufferedReader in = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
                //Now we receive the UTF-8 String sent by the client
                String str = (String)in.readLine();
                System.out.println("message= " + str);
                this.addAmount(str);

            }catch (java.net.SocketTimeoutException e){
                System.out.println("Socket timed out, restarting listening process...");
            }
            catch (IOException e) {
                e.printStackTrace();
            }

        }
    }

    public void addAmount(String amount){
        System.out.println("Placeholder method. Should be implemented in child class");
    }

    public void send_message(String message, String host, int port){
        Socket ConnectionSocket;
        try {
            ConnectionSocket = new Socket(host,port);
            //After we are connected to the socket we will send the message
            PrintWriter writer = new PrintWriter(ConnectionSocket.getOutputStream(), true);
            //We write the message we want to send in the Data Output stream
            writer.write(message);
            writer.flush();
            writer.close();
            System.out.println("Sent message:" + message + " to receiver at, IP:" + host + ", port: " + port);
            ConnectionSocket.close();
        } catch (IOException e) {
            System.out.println("Error when connecting to IP:" + host + ", port: " + port);
            e.printStackTrace();
        }

    }

    public void stop_node(){
        System.out.println("Stopping node execution");
        this.terminate_flag = 1;
    }

    public void restart_node(){
        System.out.println("Restarting node!...");
        System.out.println("Listening for requests again! IP:" + this.host + "port: " + this.port);
        this.terminate_flag = 0;
    }

    //GETTERS AND SETTERS

    public String getHost() {
        return host;
    }

    public void setHost(String host) {
        this.host = host;
    }

    public int getPort() {
        return port;
    }

    public void setPort(int port) {
        this.port = port;
    }

    public java.net.ServerSocket getServerSocket() {
        return ServerSocket;
    }

    public void setServerSocket(java.net.ServerSocket serverSocket) {
        ServerSocket = serverSocket;
    }

    public int getTerminate_flag() {
        return terminate_flag;
    }

    public void setTerminate_flag(int terminate_flag) {
        this.terminate_flag = terminate_flag;
    }
}

File: p2p-kotlin/src/Node.kt
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.*
import java.util.logging.Logger.global
import kotlin.IllegalArgumentException
import kotlin.NoSuchElementException
import kotlin.concurrent.thread

var state = 0;

open class Node(_NodeAddress : NodeAddress) {
    val address: NodeAddress = _NodeAddress
    val sock : ServerSocket
    private var running : Boolean = true;
    private var debug : Boolean = true;

    //Constructor
    init {
        //Here we will initialize the listening socket
        val endPointAddress = InetSocketAddress(InetAddress.getByName(_NodeAddress.host), _NodeAddress.port)
        sock = ServerSocket()
        //Bind the socket to the specified listening socket address (host + port)
        sock.bind(endPointAddress)
        //Set the socket timeout
        sock.soTimeout = 10000
        debugPrint("Node Created correctly! At IP: ${_NodeAddress.host} and port: ${_NodeAddress.port}\n");
    }

    fun listenConnections(): Int{
        startListening()

        thread(start = true){
            while(running){
                debugPrint("Listening for incoming connections...\n")
                //While our p2p node is running we will accept incoming connections, with each one going into a different thread
                try {
                    val ClientSocket = sock.accept();
                    debugPrint("Connection Accepted from IP: ${ClientSocket.inetAddress.hostAddress} and port: ${ClientSocket.port}\n");
                    //If the connection is accepted we will create a new thread to deal with the incomming message
                    val res = mutableListOf(0)
                    thread(start = true){
                        val result = readConnection(ClientSocket)
                        res.add(result)
                    }
                    if(res.contains(-1)){
                        print("changing state\n")
                        state = -1
                    }else if(res.contains(-2)){
                        state = -2;
                    }
                }catch (e: SocketTimeoutException){
                    debugPrint("Socket Timed out, restarting listening process...\n");
                }
            }
        }
        return 0
    }

    fun checkState(){
        debugPrint("The state is: $state\n")
        if(state == -1){
            throw NotParseableToDouble("The sent message contains some error and can't be parsed to a Double value")
        }else if (state == -2){
            throw NegativeFundsAmount("Error. Sent negative amount")
        }
        debugPrint("Returning state to normal")
        state = 0;
    }

    private fun readConnection(ClientSocket : Socket): Int {
        //read_connection deals with incoming connections
        try {
            val inputReader = BufferedReader(InputStreamReader(ClientSocket.getInputStream()));
            val message = inputReader.readLine();
            val res = receiveFunds(message);
            return res;

        }catch (e: NoSuchElementException){
            debugPrint("Error when reading the incoming message! No line was found when reading\n");
            e.printStackTrace();
            return -1
        }catch (e: NotDoubleAmount){
            debugPrint("Sent amount is not a number!");
            e.printStackTrace();
            return -1;
        }catch (e: NotParseableToDouble){

            return -1;
        }
    }

    open fun
            receiveFunds(amount : String): Int{
        debugPrint("Function to be overriden by child class");
        return -1;
    }

    fun debugPrint(text : String){
        //Simple function which prints information about the state of a Node if the variable debug is set to true
        if(debug){
            print("Debug:$text");
        }
    }

    fun sendMessage(_NodeAddress: NodeAddress, message: String): Int {
        if (_NodeAddress == this.address){
            debugPrint("Error! Trying to send a message to yourself! Check IP and port combination\n");
            val errMsg = "Can't send a message to yourself\n"
            throw SameInetAddressException(errMsg);
        }
        try {
            val outGoingSocket = Socket()
            val endPointAddress = InetSocketAddress(InetAddress.getByName(_NodeAddress.host), _NodeAddress.port)
            outGoingSocket.connect(endPointAddress);
            //Sending the message
            val output = PrintWriter(outGoingSocket.getOutputStream(), true)
            val input = BufferedReader(InputStreamReader(outGoingSocket.getInputStream()))
            output.write(message);
            output.flush();
            output.close();
            return 0;
        }catch (e: IllegalArgumentException){
            debugPrint("The port value ${_NodeAddress.port} is outside of the permitted values\n")
            e.printStackTrace();
            return -1;
        }catch (e: IOException){
            debugPrint("There was an error when connecting to IP: ${_NodeAddress.host} and port: ${_NodeAddress.port}\n")
            e.printStackTrace();
            return -1;
        }
    }

    fun stopListening(){
        debugPrint("Stopping Node from listening to messages...")
        this.running = false;
    }

    fun startListening(){
        this.running = true
    }

    fun shutdown(){
        debugPrint("Shutting down Node...")
        stopListening()
        this.sock.close();
    }


}